url49584c1d41_Arrem_GGS_tgz-pJ8-com_gamezgalaxy_test_console_MainJ8/src/com/gamezgalaxy/GGS/groups/Group.java

```diff
@@ -0,0 +1,385 @@
+	 * The permission level of this group
+	 */
+	public int permissionlevel;
+	/**
+	 * If this group is an OP group
+	 */
+	public boolean isOP = false;
+	/**
+	 * The name of this group
+	 */
+	public String name;
+	/**
+	 * The parent group
+	 */
+	public Group parent;
+	private ArrayList<String> exceptions = new ArrayList<String>(); //Commands this group can use despite permission level
+	
+	public Group(String name, int permission, boolean isOP, Group parent, Server server) {
+		this.name = name;
+		this.permissionlevel = permission;
+		this.isOP = isOP;
+		this.parent = parent;
+		server.getEventSystem().registerEvents(new Listen());
+	}
+	
+	public Group(String name, int permission, boolean isOP, Server server) {
+		this(name, permission, isOP, null, server);
+	}
+	
+	public Group(String name, Group parent, Server server) {
+		this(name, parent.permissionlevel, parent.isOP, parent, server);
+	}
+	
+	/**
+	 * Can this group execute the command <b>c</b>
+	 * This method will check the {@link #parent} group as well
+	 * if this group can't execute the command
+	 * @param c The command to check
+	 * @return Returns true if the group can execute the command
+	 */
+	public boolean canExecute(Command c) {
+		if (c.isOpCommand() && isOP)
+			return true;
+		if (c.getPermissionLevel() >= permissionlevel)
+			return true;
+		if (exceptions.contains(c.getName()))
+			return true;
+		for (String shorts : c.getShortcuts())
+			if (exceptions.contains(shorts)) {
+				return true;
+			}
+		return (parent != null) ? parent.canExecute(c) : false;
+	}
+	/**
+	 * Remove a member from this group
+	 * If you want to remove a player to this group, 
+	 * use {@link #removePlayer(Player)}
+	 * @param name The member's name name
+	 */
+	public void removeMember(String name) {
+		if (!members.contains(name))
+			return;
+		members.remove(name);
+		saveMembers();
+	}
+	/**
+	 * Add a member to this group
+	 * If you want to add a player to this group, 
+	 * use {@link #addPlayer(Player)}
+	 * @param name The member's name to add
+	 */
+	public void addMember(String name) {
+		if (members.contains(name))
+			return;
+		members.add(name);
+		saveMembers();
+	}
+	
+	/**
+	 * Add a player to this group
+	 * @param p The player to add
+	 */
+	public void addPlayer(Player p) {
+		addMember(p.username);
+		if (!online.contains(p))
+			online.add(p);
+	}
+	
+	/**
+	 * Remove a player from this group
+	 * @param p The player to remove
+	 */
+	public void removePlayer(Player p) {
+		removeMember(p.username);
+		if (online.contains(p))
+			online.remove(p);
+	}
+	
+	/**
+	 * Load the members in this group
+	 * @throws IOException
+	 *                    An IOException will be thrown if there is a problem loading the member list file
+	 */
+	public void loadMembers() throws IOException {
+		if (!new File("ranks").exists())
+			new File("ranks").mkdir();
+		if (!new File("ranks/" + name).exists())
+			return;
+		FileInputStream fstream = new FileInputStream("ranks/" + name);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fstream*. The resource is referenced at lines 155, 156, 158. The resource is closed at line 163. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *fstream* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).